srk/src/polynomialCone.ml

```diff
@@ -0,0 +1,310 @@
+                                   `Zero -> true
+                                 | _ -> false
+                               )
+                             (* use |> tap (fun intermediate -> print) *)
+                             |> BatList.of_enum
+                             |> BatList.map (fun (_, v) ->
+                                 Ideal.reduce basis (poly_of_vec v pvutil))
+                             |> BatList.filter (fun p -> not (QQXs.equal p QQXs.zero))
+  in
+  let geq_zero_constraints = Polyhedron.enum_constraints polyhedron |>
+                             BatEnum.filter (fun (constraint_kind, _) ->
+                                 match constraint_kind with
+                                   `Zero -> false
+                                 | _ -> true
+                               )
+                             |> BatEnum.map (fun (_, v) -> poly_of_vec v pvutil)
+                             |> BatList.of_enum
+  in
+  logf "finished finding %d equalities" (BatList.length equality_constraints);
+  if (BatList.length equality_constraints) > 0 then
+    (false, equality_constraints, geq_zero_constraints)
+  else
+    (true, equality_constraints, geq_zero_constraints)
+
+let rec make_enclosing_cone_impl finished basis geq_zero_polys =
+  logf "computing enclosing cone ...";
+  if finished then
+    let pc = (basis, geq_zero_polys) in
+    logf "enclosing cone: %a" pp pc;
+    pc
+  else
+    let is_finished, new_zero_polys, geq_zero_polys = find_implied_zero_polynomials geq_zero_polys basis in
+    let new_basis = BatList.fold
+        (fun ideal zero_poly -> Ideal.add_saturate ideal zero_poly)
+        basis
+        new_zero_polys
+    in
+    let reduced_geq_polys = BatList.map (fun p -> Ideal.reduce new_basis p) geq_zero_polys in
+    make_enclosing_cone_impl is_finished new_basis reduced_geq_polys
+
+let make_enclosing_cone init_ideal poly_list =
+  make_enclosing_cone_impl false init_ideal poly_list
+
+let add_polys_to_cone pc zero_polys nonneg_polys =
+  let basis, cone_generators = pc in
+  let new_basis = BatList.fold
+      (fun ideal zero_poly -> Ideal.add_saturate ideal zero_poly)
+      basis
+      zero_polys
+  in
+  make_enclosing_cone_impl false new_basis (BatList.concat [cone_generators; nonneg_polys])
+
+let project_cone cone_generators f =
+  let pvutil = pvutil_of_polys cone_generators in
+  let dim = MonomialMap.cardinal pvutil.mono_map in
+  let polyhedron_rep_of_cone = polyhedron_of_cone cone_generators pvutil in
+  let remaining_dims = (0 -- (dim - 1)) |> BatEnum.filter
+                         (fun i ->
+                            let monomial = BatList.nth pvutil.ordered_mono_list i in
+                            BatEnum.for_all
+                              (fun (d, _) -> f d)
+                              (Monomial.enum monomial)
+                         ) |> BatList.of_enum in
+  let projected_polyhedron = Polyhedron.project remaining_dims polyhedron_rep_of_cone in
```

Comment:
Polyhedron.project takes a list of dimensions to *remove* rather than the dimensions to keep.